src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,194 @@
+import * as WebSocket from 'ws';
+import { EventEmitter } from 'events';
+
+import { sbvrUtils } from '@resin/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { requestAsync } from '../../request';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = sbvrUtils;
+
+const VERSION = 1;
+
+interface LokiStream {
+	stream: {
+		device_id: string;
+		service_id: string;
+		is_system: string;
+		is_std_err: string;
+	};
+	values: Array<[string, string]>;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private webSockets: Map<string, WebSocket>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.webSockets = new Map();
+	}
+
+	public get available(): boolean {
+		// TODO: check /ready endpoint?
+		return true;
+	}
+
+	public async history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		const [response, body] = await requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/query_range`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'GET',
+			qs: {
+				query: this.getDeviceQuery(ctx),
+				limit: Number.isFinite(count) ? count : 1000,
+			},
+			json: true,
+		});
+		if (response.statusCode === 200) {
+			return this.fromLokiStreamsToDeviceLogs(body.data.result);
+		} else {
+			throw new BadRequestError(
+				`Query failed ${response.statusCode} "${response.body}"`,
```

Comment:
This message will be returned to the user, I don't know in what ways this could fail/what could be contained in the body but I'd rather `captureException` with the extra info for us to debug and send a message that makes sense for a user - avoiding exposing potentially unknown/sensitive info that might be in `response.body`

Reply:
Updated. Now `BadRequestError` returns a generic message and `captureException` is called with the error object before `BadRequest` error is thrown